File: android/app/src/main/java/edu/gatech/cs2340/rajsjarfiles/spacetrader/entity/player/Location.java
package edu.gatech.cs2340.rajsjarfiles.spacetrader.entity.player;

import edu.gatech.cs2340.rajsjarfiles.spacetrader.entity.universe.Planet;
import edu.gatech.cs2340.rajsjarfiles.spacetrader.entity.universe.SolarSystem;
import edu.gatech.cs2340.rajsjarfiles.spacetrader.model.Model;

/**
 * Represents a location in the Universe.
 */
public class Location {

    private Planet planet; // Planet that the player is currently on.
    private SolarSystem solarSystem;

    Location() {
        this(Model.getModel().getGame().getUniverse().getRandomSolarSystem());
    }

    Location(SolarSystem solarSystem) {
        this(solarSystem, solarSystem.getRandomPlanet());
    }

    Location(SolarSystem solarSystem, Planet planet) {
        setSolarSystem(solarSystem);
        setPlanet(planet);
    }

    /**
     * Sets solar system to a new solar system.
     *
     * @param solarSystem the new solar system.
     */
    private void setSolarSystem(SolarSystem solarSystem) {
        this.solarSystem = solarSystem;
    }

    /**
     * @return the solar system
     */
    public SolarSystem getSolarSystem() {
        return this.solarSystem;
    }

    public int checkIfTravelPossible(SolarSystem destinationSS, Planet destinationP) {
        // Yes possible (between planet)
        if (checkIfTravelInSS(destinationSS)) {
            return 0;
        } else {
            if (planet.getIsWarpZone()) {
                // Yes possible (between solar system)
                return 1;
            }
        }
        return -1;
    }

    private boolean checkIfTravelInSS(SolarSystem destinationSS) {
        return solarSystem.equals(destinationSS);
    }

    /**
     * Determine how much fuel is required to travel to this planet.
     *
     * @param destinationP the new planet
     * @return the amount of fuel needed
     */
    public int calculateFuelRq(Planet destinationP) {
        int distance = this.planet.getDist(destinationP);
        return distance * 5;
    }

    /**
     * Setter for planet which the player is on.
     *
     * @param planet that the player is on.
     */
    private void setPlanet(Planet planet) {
        this.planet = planet;
    }

    /**
     * @return The planet this player is currently on
     */
    public Planet getPlanet() {
        return this.planet;
    }

    /**
     * @return a list of the planets in the solar system the player
     * is in
     */
    public Planet[] getPlanets() {
        return solarSystem.getPlanets();
    }
}

File: ios/SpaceTrader/SpaceTrader/src/entity/player/Location.swift
//
//  .swift
//  SpaceTrader
//
//  Created by Thomas Suarez on 4/20/19.
//  Copyright © 2019 RajsJarFiles. All rights reserved.
//

import Foundation

class Location {
    
    var planet : Planet
    var solarSystem : SolarSystem
    
    convenience init() {
        self.init(solarSystem: Model.current.game.getUniverse().getRandomSolarSystem())
    }
    
    convenience init(solarSystem : SolarSystem) {
        self.init(solarSystem: solarSystem, planet: solarSystem.getRandomPlanet())
    }
    
    init(solarSystem : SolarSystem, planet : Planet) {
        self.setSolarSystem(solarSystem)
        self.setPlanet(planet)
    }
    
    func setSolarSystem(_ solarSystem : SolarSystem) {
        self.solarSystem = solarSystem
    }
    
    func checkIfTravelPossible(destinationSS : SolarSystem, destinationP : Planet) -> Int {
        if (self.checkIfTravelInSS(destinationSS: destinationSS)) {
            return 0
        } else {
            if (planet.isWarpZone) {
                return 1
            }
        }
        return -1
    }
    
    func checkIfTravelInSS(destinationSS : SolarSystem) -> Bool {
        return solarSystem == destinationSS
    }
    
    func calculateFuelRq(destinationP : Planet) -> Int {
        let distance = self.planet.getDist(other: destinationP)
        return distance + 5
    }
    
    func setPlanet(_ planet : Planet) {
        self.planet = planet
    }
    
    func getPlanet() -> Planet {
        return self.planet
    }
    
}
